File: paleo_workbench/ui/pages/seismic_attribute_panel.py
from __future__ import annotations

from PySide6.QtCore import Signal
from PySide6.QtGui import QBrush, QColor
from PySide6.QtWidgets import QFrame, QLabel, QTreeWidget, QTreeWidgetItem, QVBoxLayout

from paleo_workbench.seismic_attributes import available_kernels
from paleo_workbench.ui import tokens

# Kernel ids the attribute pipeline can actually compute today, with their
# user-facing labels. Anything not wired here must NOT appear as a selectable
# option — the old list advertised 17 labels with exactly 1 implementation.
_COMPUTABLE_LABELS = {
    "c3": "相干(C3)",
    "envelope": "包络",
    "rms_amplitude": "RMS振幅",
    "instantaneous_frequency": "瞬时频率",
    "instantaneous_phase": "瞬时相位",
    "sweetness": "甜点",
    "relative_impedance": "相对阻抗",
    "dip_il": "Dip_IL",
    "dip_xl": "Dip_XL",
    "dip_azimuth": "方位角",
    "curvature_mean": "平均曲率",
}

_ATTRIBUTE_GROUPS = (
    ("振幅属性", ("包络", "RMS振幅", "相对阻抗")),
    ("频率属性", ("瞬时频率", "瞬时相位", "甜点")),
    ("连续性属性", ("相干(C3)",)),
    ("构造属性", ("Dip_IL", "Dip_XL", "方位角", "平均曲率")),
    ("未实现", ("高斯曲率", "最大曲率", "RGB融合")),
)

_LABEL_TO_KERNEL = {label: kernel for kernel, label in _COMPUTABLE_LABELS.items()}


class SeismicAttributePanel(QFrame):
    """Attribute selector over the kernels the pipeline can really compute.

    Every enabled leaf maps to a wired kernel id; unimplemented reference
    labels are greyed out under an explicit 未实现 group instead of
    pretending to be options.
    """

    attribute_changed = Signal(str)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("SeismicAttributePanel")
        self.setFixedWidth(220)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
        )
        layout.setSpacing(tokens.SPACE_2)

        title = QLabel("地震属性")
        title.setObjectName("MapDockTitle")
        layout.addWidget(title)

        self.attribute_tree = QTreeWidget()
        self.attribute_tree.setObjectName("SeismicAttributeTree")
        self.attribute_tree.setHeaderHidden(True)
        self._populate_tree()
        self.attribute_tree.expandAll()
        self.attribute_tree.itemClicked.connect(self._on_item_clicked)
        layout.addWidget(self.attribute_tree, 1)

    def _populate_tree(self) -> None:
        computable = set(available_kernels())
        for group_label, labels in _ATTRIBUTE_GROUPS:
            group = QTreeWidgetItem([group_label])
            for label in labels:
                item = QTreeWidgetItem([label])
                kernel = _LABEL_TO_KERNEL.get(label)
                if kernel is None or kernel not in computable:
                    item.setDisabled(True)
                    item.setForeground(0, QBrush(QColor(tokens.TEXT_SECONDARY)))
                group.addChild(item)
            self.attribute_tree.addTopLevelItem(group)

    def _on_item_clicked(self, item: QTreeWidgetItem, _column: int) -> None:
        if item.childCount() == 0 and not item.isDisabled():
            self.attribute_changed.emit(item.text(0))

    def set_selected_attribute(self, label: str) -> None:
        """Synchronize the visual selection without triggering a new request."""
        target = str(label or "").strip()
        for group_index in range(self.attribute_tree.topLevelItemCount()):
            group = self.attribute_tree.topLevelItem(group_index)
            for child_index in range(group.childCount()):
                child = group.child(child_index)
                if child.text(0) == target:
                    self.attribute_tree.setCurrentItem(child)
                    return

    def selected_attribute(self) -> str:
        item = self.attribute_tree.currentItem()
        if item is not None and item.childCount() == 0:
            return item.text(0)
        return "振幅"

File: libs/ui_wellseis/src/qt/seismic_attribute_panel.cpp
#include <pwb/ui_wellseis/qt/seismic_attribute_panel.hpp>

#include <QTreeWidget>
#include <QTreeWidgetItem>
#include <QVBoxLayout>
#include <QLabel>

#include <pwb/ui_wellseis/seismic_attributes.hpp>

namespace pwb::ui_wellseis::qt {

namespace {

QString qs(const std::string& text) {
    return QString::fromUtf8(text.data(), static_cast<int>(text.size()));
}

const QString& unimplemented_group_label() {
    static const QString label = QStringLiteral("未实现");
    return label;
}

}  // namespace

SeismicAttributePanel::SeismicAttributePanel(
    QWidget* parent,
    std::function<bool(const std::string& kernel_id)> computable_probe)
    : QFrame(parent),
      computable_probe_(std::move(computable_probe)) {
    setObjectName(QStringLiteral("SeismicAttributePanel"));
    setMinimumWidth(200);

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(12, 12, 12, 12);
    layout->setSpacing(4);

    auto* title = new QLabel(QStringLiteral("地震属性"), this);
    title->setObjectName(QStringLiteral("MapDockTitle"));
    layout->addWidget(title);

    tree_ = new QTreeWidget(this);
    tree_->setObjectName(QStringLiteral("SeismicAttributeTree"));
    tree_->setHeaderHidden(true);
    populate();
    tree_->expandAll();
    connect(tree_, &QTreeWidget::itemClicked, this,
            [this](QTreeWidgetItem* item, int) {
                if (item == nullptr || !(item->flags() & Qt::ItemIsEnabled) ||
                    item->childCount() > 0) {
                    return;
                }
                if (!suppress_) {
                    emit attribute_changed(item->text(0));
                }
            });
    layout->addWidget(tree_, 1);
}

bool SeismicAttributePanel::enabled_leaf(const QString& label) const {
    const std::string kernel = kernel_for_label(label.toStdString());
    if (kernel.empty()) {
        return false;  // 未实现 leaf (or unknown label) — never selectable.
    }
    if (!computable_probe_) {
        return true;
    }
    return computable_probe_(kernel);
}

void SeismicAttributePanel::populate() {
    tree_->clear();
    for (const SeismicAttributeGroup& group :
         seismic_attribute_panel_groups()) {
        auto* group_item = new QTreeWidgetItem(tree_, {qs(group.group_label)});
        group_item->setFlags(Qt::ItemIsEnabled);
        for (const std::string& label : group.attributes) {
            auto* leaf = new QTreeWidgetItem(group_item, {qs(label)});
            const bool enabled = enabled_leaf(qs(label));
            leaf->setFlags(enabled ? Qt::ItemIsEnabled | Qt::ItemIsSelectable
                                   : Qt::NoItemFlags);
            if (!enabled) {
                leaf->setForeground(
                    0, tree_->palette().color(QPalette::Disabled,
                                              QPalette::Text));
                if (group.group_label == "未实现") {
                    leaf->setToolTip(0, QStringLiteral("尚未实现的参考标签"));
                }
            }
        }
    }
}

void SeismicAttributePanel::set_selected_attribute(const QString& label) {
    suppress_ = true;
    const auto items = tree_->findItems(
        label, Qt::MatchExactly | Qt::MatchRecursive);
    for (QTreeWidgetItem* item : items) {
        // Leaf-only, like the Python child scan.
        if (item != nullptr && item->childCount() == 0) {
            tree_->setCurrentItem(item);
            break;
        }
    }
    suppress_ = false;
}

QString SeismicAttributePanel::selected_attribute() const {
    QTreeWidgetItem* item = tree_->currentItem();
    if (item != nullptr && item->childCount() == 0) {
        return item->text(0);
    }
    // Python default: `return "振幅"` when no leaf is current.
    return QStringLiteral("振幅");
}

}  // namespace pwb::ui_wellseis::qt
